framework/src/base/FEProblem.C

```diff
@@ -3837,6 +3875,9 @@ FEProblem::checkNonlinearConvergence(std::string &msg,
                                      const Real initial_residual_before_preset_bcs,
                                      const Real div_threshold)
 {
+  if (_has_exception)
```

Comment:
This won't really work, right?  Because you're clearing `_has_exception` when you handle it.

I think you just need to check the PETSc reason and if it's `DIVERGED_DOMAIN_ERROR` then set this.